RFD 219 - Kubernetes RBAC simplification

creack opened the pull request:
Re-open with proper branch name and updated number from #51707

RFD to simplify Kubernetes RBAC. Based on prior work around CRD support, Namespace/Kind behavior change in role v8 and complete resource registrations.

Focus on removing the need for `kubernetes_user` and `kubernetes_group`.

creack committed:
RFD Proposal for K8s RBAC simplification.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

rosstimothy reviewed (commented):
What impact will this change have on trusted clusters?

creack commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,260 @@
+
+Role v8 also changed how the wildcard kind works, it now enforces the
+`namespace` field, which allows setting permission for a specific cluster-wide
+or namespaced resource easily.
```
It was just to highlight was has already been done related to the RFD, but indeed, already merged. Removed the prior work section.

hugoShaka reviewed (changes requested):
This RFD is suggesting a breaking change that will require the users to manually update/redeploy every kubernetes agent they have, such RBAC change cannot be pushed via autoupdate.

The existing users will not gain capabilities from this change so they will likely never do the migration. We already experienced how difficult it is to get cloud customers to enable automatic updates, the last thing we want to do is have to manually contact cloud users again so they can have agent supporting the new role version.

Without those infra changes, users will not be able to use the features described in the rest of the RFD (role v9). This will cause user base fragmentation, existing users won't be able to follow docs anymore (as docs will tell people to create rolev9). This will increase support cost for us and generate frustration for our existing users.

The issue will be worse for large customers who have different teams but might be sharing roles. Even if they want to use rolev9, they won't be able to until all of their teams perform the infra changes, updating can take months.

We cannot afford to break existing users for the sake of improving the onboarding experience. Teleport already struggles to keep its users up-to-date. Such a change makes updating harder and goes against what we've been trying to fix for the past year. I think the drawbacks are several orders of magnitude above the benefits.

creack commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,325 @@
+We'll need to update the Helm chart, the `get-kubeconfig.sh` provisioning
+script as well as the discovery service for ASK, EKS, and GCP.
+
+The new Cluster Role will be named `teleport-cluster-admin` and will have the
```
Added a mention to refer to the Configuration section.

creack commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,325 @@
+name, we need to add a new field in the kube_service configuration file to
+define which admin group will be impersonated.
+
+This can also be used for customer upgrading from older versions that don't
+want to provision a new clusterrole/clusterrolebinding. The config value could
+be set to `cluster-admin` or `system:masters` to remove the need to provision
+new resources.
```
Done.

creack commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,325 @@
+
+#### Helm Config
+
+The Helm Chart will provide the ability to specify custom names for all
+resources, (clusterrole, clusterrolebinding) and for the Group.
```
Done.

creack commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,252 @@
+where this group is forbidden).
+Teleport is super admin and yields reduced permissions to the user based on
+the `kubernetes_resources` and `kubernetes_label` fields from the Teleport
+role.
```
Updated. Thanks for the clarification.

klizhentas commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,197 @@
+
+## Goals
+
+- Discourage the use of `kubernetes_groups` and `kubernetes_users` fields from
```
you mean discourage for getting started, not in general?

klizhentas commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,197 @@
+created by provisioning scripts or auto-discovery.
+This will allow users to avoid the complexity of managing RBAC in multiple
+places, avoid confusion around the `exec` subresource.
+While we won't change or remove the `kubernetes_groups` and `kubernetes_users`
```
what should they use instead for fine-grained access?

creack committed:
Add section about user flow for day 0 / day 1.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

klizhentas commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,242 @@
+    ClusterRole. The `kubernetes_groups` field is pre-populated in the Web UI
+    and well documented in the Teleport documentation for the YAML version.
+- Day 1 - Ongoing management:
+  - User can manage the Teleport role without needing to understand the
```
if that's the case, what is the purpose of `kubernetes_groups` in the preset role `kube-editor`?

creack replied:
While it could be removed for new customers, it's purpose it to maintain compatibility with existing customers who may not be willing to apply the new clusterrole/clusterrolebinding to enable the simplification.

creack commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,242 @@
+    or Kubernetes).
+  - User may need to troubleshoot unexpected permissions or access issues.
+
+#### Proposed flow
```
We create a new preset to make day 0 easier as there is currently no preset roles granting kubernetes access.
The `kubernetes_groups` field is required and allows Teleport itself to have cluster-admin permissions, allowing us to be authoritative and let the user only worry about labels and resources instead.

creack committed:
PR feedback - add conifguration field, use wildcard for groups.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

creack committed:
Add section about downgrade in rfd.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

creack committed:
Update RFD with latest consensus/feedback.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

creack committed:
Update RFD to use presets instead of managing clusterroles manually.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

creack committed:
fix typo.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

creack committed:
PR feedback.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

creack committed:
Add rollout section.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

creack committed:
Add backport section.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

russjones commented:
Waiting for one last review from @hugoShaka before we merge this.

tigrato reviewed (commented):
_ignore if you want _

my2c

I don’t think this really addresses the core issue Teleport users run into. Access is still fragmented and governed by per-cluster Kubernetes RBAC.

For example, if a cluster was enrolled before this feature, the preset bindings won’t be there. But if another cluster is enrolled after the feature, the presets will exist. A user from the second cluster trying to access the first cluster will be denied - even if from Teleport’s perspective everything looks correct and the agent versions match - simply because the first cluster was installed with an older Helm chart. The result is just a generic “access denied” from Kubernetes with no clear error from Teleport -  teleport has no clear way of checking if a group exists and if it should exist in the first place.

#39628 was meant to fix this by ensuring roles are defined centrally and always synced across clusters. This remains a common `c-na` request that users frequently bring up.

This is helpful for day-0 setup, but granting system:masters would make things even simpler because users know what `system:masters` means but don't know what `teleport:preset:editor` means exactly. 

From my experience, most users don’t rely on Kubernetes’ built-in edit or view roles, since they typically manage access through per-namespace RBAC instead

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,370 @@
+
+#### New Teleport preset roles
+
+In order to be consistent and allow for seamless use of Kubernetes within
+Teleport, we will introduce the following new preset roles:
+
+- `kube-access`: Maps to the Kubernetes preset `edit`. Can CRUD most basic
+  builtin resources and read some limited administrative resources.
+  This would be used in conjunction with the `access` Teleport role to grant
+  most non-administrative access to users.
+- `kube-editor`: Maps to the Kubernetes preset `cluster-admin`. Provides full
+  access to all Kubernetes resources, including CRDs.
+  This would be used in conjunction with the `editor` Teleport role to grant
+  full administrative access to users.
+- `kube-auditor`: Maps the the Kubernetes preset `view`. Provides read-only
+  access to some Kubernetes resources.
+  This would be used in conjunction with the `auditor` Teleport role to grant
+  read-only access to users.
```
As this section is discussing consistency, I wonder if we should be consistent with the way we manage day-0 SSH access today.

We have the `access` role with a wildcard selector and we use the `logins` user traits to grant access to specific users. Database and windows access also work the same way with `database_users` and `windows_logins`.

Once the Kubernetes Bindings described in this RFD are created, we can provide access with the existing `access` preset and user traits, without creating new roles.

This approach would have the following advantages:
- consistency with the other protocols which don't have their dedicated presets
- reduce the probability of users shooting themselves in the foot with the awkward kube_group + kube_resoucres interactions
- leaner and reduces the number of Teleport presets

The guides would tell day 0 users to do:

```bash
tctl users add hugo --logins root --kubernetes-group teleport:preset:editor --roles=access
```

And they would have access by default.

creack replied:
That's a good point. I updated the RFD to reflect this. Now we don't create any preset on the Teleport side but focus on the Kubernetes provisionning. Creating the bindings as described before and rely on the user trait to interpolate which K8s role will be used.

cc @klizhentas this simplifies even further but significantly changes the approved proposal, can you take another look?

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,370 @@
+role.
+
+- `kubernetes_groups` -> `ClusterRole`
+- `teleport:preset:access` -> `edit`
+- `teleport:preset:editor` -> `cluster-admin`
+- `teleport:preset:auditor` -> `view`
```
nit: I'd stick to kubernetes cluster role names in the CRB so it's clear what they bind to. Users might want to reuse them.

```suggestion
- `teleport:preset:edit` -> `edit`
- `teleport:preset:cluster-admin` -> `cluster-admin`
- `teleport:preset:view` -> `view`
```

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
````diff
@@ -0,0 +1,280 @@
+  auditorGroupName: teleport:preset:auditor
+```
+
+Note that if a user decides to change the preset names, they will not be able
+to use the preset Teleport roles out of the box, as they will need to update
+them to match the new names.
````
> if we fail to create the new teleport:preset resources in Kubernetes?

The chat won't deploy unless you disable the RBAC creation, and in this case I suppose that you want to handle the RBAC on your own and not user what we provide.

If you want you can also create your own rolebindings for our standard group.

creack committed:
Address PR feedback.

Signed-off-by: Tim Ross <tim.ross@goteleport.com>

rosstimothy committed:
Simplify design

Removes the ability to set custom names for ClusterRoleBindings and
instead provides a boolean flag to disable provisiong of Kubernetes
RBAC resource automatically.

Removes left over references to changing the default `kubernetes_groups`
values to prese bindings. All roles will be left referencing the
internal traits and all documentation will be updated to specify
the preset roles as traits for day 0 guides.

rosstimothy committed:
Detail CRB prefixing

rosstimothy committed:
fix: spelling

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,280 @@
+role.
+
+Those bindings will be available to Teleport users via the user trait feature.
+
```
Adding more details from the discussions with @klizhentas about why we prefer to not do this ourselves.

```suggestion

Kubernetes might introduce new API endpoints that should be allowed or denied.
Vendoring those roles on Teleport would require us to update the list for every
Kubernetes version, and would potentially cause a permission issues if a user
deploys an old Teleport version in a modern Kubernetes cluster. Relying on the
Kubernetes built-in roles for RBAC frees us from the maintenance burden and
prevents security issues caused by version mismatch.

```

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,280 @@
+
+### Provisioning
+
+To enable easy setup, provisioning methods will be updated to create the
+required new ClusterRoles and ClusterRoleBindings.
```
We use the kube built-in ClusterRoles, no need to create them:

```suggestion
To enable easy setup, provisioning methods will be updated to create the
required new ClusterRoleBindings.
```

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,280 @@
+{{- if .Values.rbac.defaultClusterRoleBindings -}}
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:metadata:
```
```suggestion
metadata:
```

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,280 @@
+
+#### Provision Script
+
+For simplicity, the provision script will be updated to always create the
+Kubernetes resources, and will not provide a means for users to opt out of
+the new behavior.
```
Not sure which script we are talking about

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,280 @@
+The following page is about testing the connection, and also have a form to
+populate a subset of the users/groups.
+
+The initial idea was to allow the user to update it's traits on the fly to not
+have to logout/log back in, however, as there is no mention of traits, and
+because the form is present on multiple pages with different behavior, and
+because it requires the user to initially already have a role assigned that
+contains the trait interpolation, we will remove both the 'Setup access' page
+and the 'users/groups' form from the 'Test connection' page.
```
We might need to leave them for existing Teleport users who don't have the kube_user and kube_group traits. We can populate them by default though. @marcoandredinis is the owner of the discovery UI process and might have opinions on this.

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,280 @@
+
+#### Proposed flow
+
+- Day 0 - Initial setup:
+  - User installs Teleport and sets up a Kubernetes cluster.
+  - User provisions the Kubernetes cluster using a script, discovery, or Helm
+    chart to create all required RBAC resources.
+  - User configures the Teleport role with `kubernetes_resources` and
+    `kubernetes_labels` to match or reduce the permissions granted by the
+    ClusterRole.
+  - User applies the desired `kubernetes_groups` trait to the Teleport user.
+- Day 1 - Ongoing management:
+  - User can reduce the scope of either Kubernetes or Teleport's roles by
+    changing the counterpart
+  - User has the option to use the `teleport:preset:editor` role to avoid the
+    complexity of managing the underlying Kubernetes RBAC, only managing
+    resources/labels on the Teleport side instead.
```
```suggestion
- Day 0 - Initial setup, prospect and hackers:
  - Prospect installs Teleport
  - Prospect creates its first Teleport local user with the trait `kubernetes_groups: "teleport:preset:cluster-admin"`
  - Prospect deploys a Teleport agent in their Kubernetes cluster
  - Prospect can access the Kubernetes cluster as administrator with their local Teleport user
- Day 1 - Multiple users, late stage prospect or new Teleport adopter
  - Admin has some Kubernetes agents connected, SSO connector set up, and several users
  - Admin creates per-team roles with different `spec.allow. kubernetes_labels` to select each team's cluster
  - Admin sets the `teleport:preset:edit` or `teleport:preset:view` as `spec.allow.kubernetes_groups`
  - Admin grants the role to users using SSO connector or Access Lists
  - users from each team can access their Kubernetes clusters
- Day 2 - More granular access, advanced Teleport users
  - Admin can create custom Kubernetes rolebindings, deploy them in their clusters and use them in their Teleport roles instead of the presets
  - Admin can scope down read and edit roles to select namespaces or resources by using `kubernmetes_resources` in their roles
```

hugoShaka commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,280 @@
+
+### UX
+
+#### Exec confusion
+
+As the underlying Kubernetes RBAC gets discouraged, it will help with the
+confusion around _exec_ being allowed with a `get` verb is gone.
+Teleport uses a dedicated verb to control access to `exec`.
+
```
```suggestion
```

rosstimothy committed:
fixup: remove reference to provisioning ClusterRoles

Co-authored-by: Hugo Shaka <hugo.hervieux@goteleport.com>

rosstimothy committed:
fix: Remove exec confusion section

Co-authored-by: Hugo Shaka <hugo.hervieux@goteleport.com>

rosstimothy committed:
fix: describe proposed day 0,1,2 flows better

Co-authored-by: Hugo Shaka <hugo.hervieux@goteleport.com>

rosstimothy committed:
fix: remove duplicate metadata:

Co-authored-by: Hugo Shaka <hugo.hervieux@goteleport.com>

rosstimothy committed:
Update required approvers

rosstimothy committed:
Update RFD authors

rosstimothy committed:
fix: typo

rosstimothy committed:
Apply suggestion from @hugoShaka

Co-authored-by: Hugo Shaka <hugo.hervieux@goteleport.com>

rosstimothy committed:
fix: clarify what the provisioning script is

r0mant reviewed (approved):
lgtm with one optional comment for consideration but no blockers from me as long as this has been reviewed by product

r0mant commented on rfd/0219-k8s-rbac-simplification.md:
```diff
@@ -0,0 +1,281 @@
+For consistency and reliability, we will create one binding for each preset
+role.
+
+Those bindings will be available to Teleport users via the user trait feature.
```
I see now that it was [discussed above](https://github.com/gravitational/teleport/pull/56214#discussion_r2310361102). TBH I don't see a problem or a consistency issue with including more preset roles if it improves user experience (which I do think it does compared to having to assign traits) - e.g. AWS has literally hundreds of builtin/preset roles for each imaginable scenario which makes it easy to find and pick one you need (or use as a base to tweak policies). In fact, those specialized roles I think would actually be better from security standpoint than relying on blanket access role as well, and if somebody wants to do this via traits, they can too. But like mentioned above I don't think it's a blocker.
